### PR TITLE
fix #10637 assign provider namespace

### LINKED ISSUE #10637: dubbo mesh开启后,如何跨namespace调用与dubbo多版本调用
- [x] I have searched the [issues](https://github.com/apache/dubbo/issues) of this repository and believe that this is not a duplicate.

## Ask your question here
操作步骤如下：
1. k8s环境下，consumer端开启mesh模式
2. consumer端需要调用多个provider
3. provider有应用级注册和接口级注册，并且需要的provider不再同一个namespace下
4. 原有项目存在dubbo2.6与3.1多版本间的相互调用，在开启消费端mesh（dubbo3.1）后，无法调用到2.6的服务提供者

**问题：**

1. 如何实现在一个consumer下，调用多个namespace下的provider，并且provider有应用级注册和接口级注册
2. dubbo3.1消费端在开启mesh后，如何调用版本是2.6的服务提供者

### PR DESCRIPTION
**Pull request by niejian**
## What is the purpose of the change
add attribute(providerNamespace) in @DubboReference, and add `provider-namespace` attribute in xsd, that fix issue https://github.com/apache/dubbo/issues/10637

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [ ] GitHub Actions works fine on your own branch.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Comment by conghuhu** [comment]
LGTM
@AlbumenJ

File: dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ReferenceConfig.java [hunk 1]
```diff
@@ -478,16 +478,21 @@ private void meshModeHandleUrl(Map<String, String> referenceParameters) {
...
         }
-        // get pod namespace
-        String podNamespace;
-        if (StringUtils.isEmpty(System.getenv("POD_NAMESPACE"))) {
-            if (logger.isWarnEnabled()) {
-                logger.warn(CONFIG_FAILED_LOAD_ENV_VARIABLE, "", "", "Can not get env variable: POD_NAMESPACE, it may not be running in the K8S environment , " +
-                    "finally use 'default' replace.");
+
+        // get provider namespace if @DubboReference present
+        String podNamespace = Optional.ofNullable(referenceParameters.get(RegistryConstants.PROVIDER_NAMESPACE)).orElse(null);
```
Comment (AlbumenJ) [inline dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ReferenceConfig.java:483]: why use `Optional.ofNullable`

Comment (AlbumenJ) [inline dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ReferenceConfig.java:483]: why use `Optional.ofNullable`

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ReferenceConfig.java (hunk 1 above)
Comment (niejian) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ReferenceConfig.java:483]: avoid npe

Comment (AlbumenJ) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ReferenceConfig.java:483]: `Optional.ofNullable(referenceParameters.get(RegistryConstants.PROVIDER_NAMESPACE))` could not avoid NPE when referenceParameters is null

Comment (niejian) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ReferenceConfig.java:483]: @AlbumenJ 

改成这样？

image

**Comment by niejian** [comment]
@AlbumenJ 
I have alread changed the way to get  `namespace` -> `podNamespace = referenceParameters.getOrDefault`. Pls check it

File: dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ReferenceConfig.java (hunk 1 above)
Comment (AlbumenJ) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ReferenceConfig.java]: 直接 get 就行了呀，默认空就是返回 null

另外下面 `null == podNamespace` 最好换成 StringUtils.isNotEmpty

Comment (niejian) [reply dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ReferenceConfig.java]: 已经改了

**Pull request merged**